src/main.cpp

```diff
@@ -3838,11 +3841,33 @@ bool ReceivedBlockTransactions(
         deque<CBlockIndex*> queue;
         queue.push_back(pindexNew);
 
+        CCoinsViewCache view(pcoinsTip);
+
         // Recursively process any descendant blocks that now may be eligible to be connected.
         while (!queue.empty()) {
             CBlockIndex *pindex = queue.front();
             queue.pop_front();
             pindex->nChainTx = (pindex->pprev ? pindex->pprev->nChainTx : 0) + pindex->nTx;
+
+            // Given that all parents are BLOCK_VALID_TRANSACTIONS, we now should be
+            // able to calculate this block's net transparent value, assuming the
+            // block is correct.
+            CAmount transparentValue = 0;
+            for (auto& tx : block.vtx) {
+                for (auto& out : tx.vout) {
+                    transparentValue += out.nValue;
+                }
+                if (view.HaveInputs(tx)) {
+                    transparentValue -= view.GetTransparentValueIn(tx);
+                } else {
+                    // Some of the inputs are inaccessible; this might be a block from
+                    // a fork that spends coins also spent in the current chain. Fall
+                    // back to directly looking up the transactions.
+                    // TODO
```

Comment:
Do you have some more input on how we will implement this ? Thanks.

Reply:
If we can ensure that we only call this in-order, then we can make that we have the inputs an assertion. Changing the enclosing procedure to take the coins view as an argument will help clarify whether we can do this; otherwise we'd have to look up the UTXOs from the txindex if present.